### PR TITLE
Audience checking

### PR DESCRIPTION
**Pull request by mstruk**
Support for audience checking.

When enabled the Kafka broker rejects during authentication any access token that does not contain the Kafka broker's OAuth client ID in`aud` claim.

### CODE REVIEW

**Comment by tombentley** [comment]
@mstruk any links to documentation about what "audience check" _is_?

**Comment by mstruk** [comment]
> @mstruk any links to documentation about what "audience check" _is_?

I don't know about a simple nice explanation.
It's mentioned in several places in the [OpenID Connect spec (extension of OAuth2)](https://openid.net/specs/openid-connect-core-1_0.html#IDToken).

**Comment by scholzj** [comment]
> I don't know about a simple nice explanation.
It's mentioned in several places in the OpenID Connect spec (extension of OAuth2).

It will anyway need to be documented. So I guess if there is no reasonable explanation what this does we (well, you) would need to write it.

File: README.md
Comment (scholzj) [inline README.md:271]: So, the JWT token contains the `aud` field, right? And the `oauth.check.audience` says that it should be checked. But checked against what? Where do you set the allowed `aud` values?

Comment (mstruk) [reply README.md:271]: The valid value is only one - the configured `oauth.client.id`. Maybe it's not adequately explained in the README.md?
For example - if client id is configured as 'kafka' which it usually is, then the examples of valid `aud` value are:

* "kafka"
* ["rest-api", "kafka"]

**Comment by mstruk** [comment]
> It will anyway need to be documented. So I guess if there is no reasonable explanation what this does we (well, you) would need to write it.
The additional option is documented in README.md, but maybe not sufficiently. I'll add a sentence or two..

File: README.md
Comment (tombentley) [inline README.md]: I think "limit access" is a bit confusing. IIUC it's not that there is something which can't _access_ the tokens, it's that the tokens are rejected as invalid because they weren't issued to be used for that Kafka broker (acting in the role of resource server). Maybe "...which is an OAuth2 mechanism to prevent validation of tokens that weren't explicitly issued for use by your protected server." would be better phrasing.

Also, normal English usage would be "You can enforce audience checking".

File: README.md
Comment (tombentley) [inline README.md]:
```suggestion
Set the following option to `true` to enforce audience checking (it is not enabled by default). 
```

File: README.md
Comment (tombentley) [inline README.md:274]:
```suggestion
When audience checking is enabled the `oauth.client.id` has to be configured:
```

File: README.md
Comment (tombentley) [inline README.md:275]:
```suggestion
- `oauth.client.id` (e.g.: "kafka" - this is the OAuth2 client configuration id for the Kafka Broker)
```

File: README.md
Comment (tombentley) [inline README.md:327]:
```suggestion
- `oauth.client.id` (e.g.: "kafka" - this is the OAuth2 client configuration id for the Kafka Broker)
```

File: README.md
Comment (tombentley) [inline README.md]: Same comment

(also why are we repeating this verbatim , can't we just refer the reader to the previous duscussion?)

**Comment by mstruk** [comment]
> It's mentioned in several places in the [OpenID Connect spec (extension of OAuth2)](https://openid.net/specs/openid-connect-core-1_0.html#IDToken).

[Another link](https://tools.ietf.org/html/draft-ietf-oauth-json-web-token-25#section-4.1.3) (for reference).

File: oauth-common/src/main/java/io/strimzi/kafka/oauth/validator/OAuthIntrospectionValidator.java
Comment (tombentley) [inline oauth-common/src/main/java/io/strimzi/kafka/oauth/validator/OAuthIntrospectionValidator.java:216]: When `value` is a JSON string, rather than a JSON array of strings, `JSONUtil.asListOfString` will split it on whitespace, which doesn't seem to be correct according to https://tools.ietf.org/html/draft-ietf-oauth-json-web-token-25 or https://openid.net/specs/openid-connect-core-1_0.html#IDToken. It _might_ be, because "The interpretation of audience values is generally application specific.". 

The JWT document also says that the content of the JSON strings can also be a URI if the String contains a `:`, Is that something we need to think about here?

File: README.md
Comment (mstruk) [reply README.md]: OTOH when I hear 'enforce audience checking' I specifically think of this functionality being enabled vs. not being enabled. I don't think about the process of validating the token being triggered and failing. Rather I think about the process of validating not being attempted at all.

Comment (mstruk) [reply README.md]: Ah, sorry I misunderstood your last sentence. So with regards to 'limit access', what about saying it like this:

"...which is an OAuth2 mechanism to prevent *authentication* of tokens that weren't explicitly issued for use by your protected server."

File: README.md
Comment (mstruk) [reply README.md]: It's structured to have 'fast local JWT checking' chapter with all the config options, and a separate 'introspection based validation' chapter. Some options are shared by both. But if you look closely you'll see that there are some nuances and even the part about audience checking is slightly different for 'fast JWT checking' chapter. There's this sentence:

"When audience check is enabled the `oauth.client.id` has to be configured:
- `oauth.client.id` (e.g.: "kafka" - this is the OAuth2 client configuration id for Kafka Broker)" 

Which is not present in 'Introspection based validation' chapter, because `oauth.client.id` is already a required configuration there.

I prefer it this way as it's more user-friendly than having the user jump around and trying to figure out if something fits together with something else.

File: oauth-common/src/main/java/io/strimzi/kafka/oauth/validator/OAuthIntrospectionValidator.java
Comment (mstruk) [reply oauth-common/src/main/java/io/strimzi/kafka/oauth/validator/OAuthIntrospectionValidator.java:216]: Good catch about `JSONUtil.asListOfString()` even though I think it behaves slightly different than you described :)

If the value is `"kafka"` then it's a text node, and it will be correctly parsed. 
If the value is `"kafka rest-api"` - that's the application specific handling, and not by spec - it will be correctly parsed under the assumption that the client id can't contain a space character (I have not checked this assumption). 
If the value is `"kafka, rest-api"` it is a text node again and not by the spec - it WON'T be correctly parsed unless there is client id containing a comma and a space (I assume these are not valid characters for client id). I also assume this case would never happen that multiple client ids would be sent this way - as a single string.
If the value is `["kafka", "rest-api"]` or `["kafka"]` it is an array node and will be handled correctly.

We could improve the JSONUtil.asListOfString() where it parses the list of strings sent as a single string to also treat the comma as a separator, not just the space. Assuming that all the contexts where it is used don't allow comma and space as legal characters inside individual string items or that no reasonable person would use these characters in client ID.

The spec indeed says that it could be a url, but I'm unaware of any implementation using that, and if we go testing proper support if url is the value then we're possibly spending time on something that no-one would ever use and that has no real implementation for us to test with.

File: README.md
Comment (mstruk) [reply README.md]: Or maybe it should be:

"...which is an OAuth2 mechanism to prevent *successful authentication with* tokens that weren't explicitly issued for use by your protected server."

Comment (tombentley) [reply README.md]: SGTM.

File: oauth-common/src/main/java/io/strimzi/kafka/oauth/validator/OAuthIntrospectionValidator.java
Comment (tombentley) [reply oauth-common/src/main/java/io/strimzi/kafka/oauth/validator/OAuthIntrospectionValidator.java:216]: The spec seems to be silent about whether the string can contain whitespace or not. I assumed they could, but that such values should not be treated as a list of tokens, but rather a single value containing a string. In any case I'll defer to you on this, I just wanted to point this stuff out.

File: README.md
Comment (tombentley) [inline README.md:268]:
```suggestion
You can enforce audience checking, which is an OAuth2 mechanism to prevent successful authentication with tokens that are not explicitly issued for use by your resource server.
```

File: README.md
Comment (tombentley) [inline README.md:335]:
```suggestion
You can enforce audience checking, which is an OAuth2 mechanism to prevent successful authentication with tokens that are not explicitly issued for use by your resource server.
```

File: README.md
Comment (tombentley) [inline README.md:336]:
```suggestion
The authorization server adds the allowed resource servers' `client IDs` into the `aud` claim of such tokens.
```

File: README.md
Comment (tombentley) [inline README.md:269]:
```suggestion
The authorization server adds the allowed resource servers' `client IDs` into the `aud` claim of such tokens.
```

**Review by tombentley — APPROVED** [review]

File: README.md
Comment (mstruk) [reply README.md:268]: Good catch, I suggest a slightly different fix to keep consistency with the next sentence.
```suggestion
You can enforce audience checking, which is an OAuth2 mechanism to prevent successful authentication unless the tokens are explicitly issued for use by your resource server.
```

File: README.md
Comment (mstruk) [reply README.md:335]: Again ...
```suggestion
You can enforce audience checking, which is an OAuth2 mechanism to prevent successful authentication unless tokens are explicitly issued for use by your resource server.
```

**Review by scholzj — APPROVED** [review]

**Comment by scholzj** [comment]
@mstruk Looks good to me ... but you will need to fix the DCO signoff.

**Comment by scholzj** [comment]
And the CI build I guess.

File: oauth-common/src/main/java/io/strimzi/kafka/oauth/validator/OAuthIntrospectionValidator.java
Comment (mstruk) [reply oauth-common/src/main/java/io/strimzi/kafka/oauth/validator/OAuthIntrospectionValidator.java:216]: Let's just leave it like this for now.

File: README.md
Comment (mstruk) [reply README.md]: Slightly reworded to fit with the line that follows:

"You can enforce audience checking, which is an OAuth2 mechanism to prevent *successful authentication with* tokens **unless they are** explicitly issued for use by your resource server."

**Comment by mstruk** [comment]
Testsuite passes, DCO fixed. AFAIC this is good to merge.

**Pull request merged**
